py/htm/encoders/eye.py

```diff
@@ -1,27 +1,119 @@
 # Written by David McDougall, 2018
 
-""" From Wikipedia "Retina": Although there are more than 130 million
+"""
+From Wikipedia "Retina": Although there are more than 130 million
 retinal receptors, there are only approximately 1.2 million fibres
 (axons) in the optic nerve; a large amount of pre-processing is
 performed within the retina. The fovea produces the most accurate
 information. Despite occupying about 0.01% of the visual field (less
 than 2 of visual angle), about 10% of axons in the optic nerve are
 devoted to the fovea.
 
+For more information on visual processing on retina, see 
+https://foundationsofvision.stanford.edu/chapter-5-the-retinal-representation/#visualinformation
+
+
 Fun Fact 1: The human optic nerve has 800,000 ~ 1,700,000 nerve fibers.
 Fun Fact 2: The human eye can distiguish between 10 million different colors.
-Sources: Wikipedia. """
+~Sources: Wikipedia. 
+"""
 
-import numpy as np
-import cv2
-import scipy.misc
 import math
-import scipy.ndimage
 import random
-import PIL, PIL.ImageDraw
-import matplotlib.pyplot as plt
-from sdr import SDR
-import encoders
+
+import numpy as np
+import cv2 # pip install opencv-contrib-python
+from PIL import Image # pip Pillow
+
+from htm.bindings.sdr import SDR
+
+
+class ChannelEncoder:
+    """
+    This assigns a random range to each bit of the output SDR.  Each bit becomes
+    active if its corresponding input falls in its range.  By using random
```

Comment:
I wrote at length about this on the Numenta Forum.  I'm going to copy paste it all here, for us to edit and eventually move into the documentation.

Two requirements for encoding SDRs are that every bit of information represents a range of possible values, and that for every input multiple bits activate in the SDR. The effects of these requirements are that each output bit is inaccurate and redundant. This encoder makes every output bit receptive to a unique range of inputs, which are uniformly distributed throughout the input space. This meets the requirements for being an SDR and has more representational power than if many of the bits represented the identical input ranges. This design makes all of those redundancies add useful information.

Semantic similarity happens when two inputs which are similar have similar SDR representations. This encoder design does two things to cause semantic similarity: (1) SDR bits are responsive to a range of input values, and (2) topology allows near by bits to represent similar things.

1) Many encoders apply thresholds to real valued input data to convert the input into Boolean outputs. In this encoder uses two thresholds to form ranges which are referred to as ‘bins’. A small change in the input value might cause some of the output bits to change and a large change in input value will cause all of the output bits to change.  How sensitive the output bits are to changes in the input value -the semantic similarity- is determined by the sizes of the bins. The sizes of the bins are in turn determined by the sparsity, as follows:
    Assume that the inputs are distributed in a uniform random way throughout the input range. The size of the bins then determines the probability that an input value will fall inside of a bin. This means that the sparsity is related to the size of the bins, which in turn means that the sparsity is related to the amount of semantic similarity. This may seem counter-intuitive but this same property holds true for all encoders which use bins to convert real numbers into discrete bits.

2) This encoder relies on topology in the input image, the idea that adjacent pixels in the input image are likely to show the same thing.  If an area of the output SDR can not represent a color because it did not generate the bins needed to, then it may still be near to a pixel which does represent the color. In the case where there are less than one active output bits per input pixel, multiple close together outputs can work together to represent the input.  Also, if an output bit changes in response to a small change in input, then some semantic similarity is lost. Topology allows nearby outputs to represent the same thing, and since each output uses random bins they will not all change when the input reaches a single threshold.

Note about sparsity and color channels:
To encode color images create separate encoders for each color channel. Then recombine the output SDRs into a single monolithic SDR by multiplying them together. Multiplication is equivalent to logical “and” in this situation. Notice that the combined SDR’s sparsity is the different; the fraction of bits which are active in the combined SDR is the product of the fraction of the bits which are active in all input SDRs. For example, to make an encoder with 8 bits per pixel and a sparsity of 1/8: create three encoders with 8 bits per pixel and a sparsity of 1/2.

Reply:
Thanks, David :+1: 
I'll study this later on,

> This encoder makes every output bit receptive to a unique range of inputs, which are uniformly distributed throughout the input space. 

so a uniformly random subsample (of fixed size?) of bits is selected as receptive field for each output bit (?) 

- Numenta used to have a "RandomSP" mode, which was a SP without learning, only the initial random weights were used. Wouldn't that suffice the described behavior? 

> (2) topology allows near by bits to represent similar things. 

I haven't noticed any topology used in ChEnc? 

> This encoder relies on topology in the input image, the idea that adjacent pixels in the input image are likely to show the same thing.  

ok.. 

Overall: 

- this does seem like a useful way for a (generic) encoder
- [ ] can we use RDSE instead for this PR, and deal with `ChannelEncoder` separately? 
  - I somehow thought it's designed to deal with a stream of images, but what you describe sounds like an approach to a generic encoder

Reply:
> can we use RDSE instead for this PR, and deal with ChannelEncoder separately? 

No, because the channel encoder can encode each input scalar (pixel) into as few as 1 bit, but I think that the RDSE would break if told to encode an input scalar into 1 bit.